src/symbols.c

```diff
@@ -1597,13 +1597,24 @@ static TMTag *find_best_goto_tag(GeanyDocument *doc, GPtrArray *tags)
 
 static GPtrArray *filter_tags(GPtrArray *tags, TMTag *current_tag, gboolean definition)
 {
+	GeanyDocument *doc = document_get_current();
+	guint current_line = sci_get_current_line(doc->editor->sci) + 1;
 	const TMTagType forward_types = tm_tag_prototype_t | tm_tag_externvar_t;
 	TMTag *tmtag, *last_tag = NULL;
+	const gchar *current_scope = NULL;
 	GPtrArray *filtered_tags = g_ptr_array_new();
 	guint i;
 
+	symbols_get_current_function(doc, &current_scope);
+
 	foreach_ptr_array(tmtag, i, tags)
 	{
+		if (tmtag->type & tm_tag_local_var_t &&
+			(doc->tm_file != tmtag->file ||
+			 current_line < tmtag->line ||
```

Comment:
`tm_parser_var_valid_before_declare()` should be called here, no?

Also, the whole condition is basically a duplicate of the same in `is_valid_autocomplete_tag()`. Can `is_valid_autocomplete_tag()` be made callable from here (perhaps with a different name)?

Reply:
> tm_parser_var_valid_before_declare() should be called here, no?

In fact, I'm less convinced whether `tm_parser_var_valid_before_declare()` is a good idea now - even though you can declare (i.e. assign) variables after their first appearance in file, I think this pattern is not so common and like in the C/C++ case because of this you will get invalid variables in the autocompletion in many cases which IMO is not worth it.

> Also, the whole condition is basically a duplicate of the same in is_valid_autocomplete_tag(). Can is_valid_autocomplete_tag() be made callable from here (perhaps with a different name)?

I was thinking about it but there's small difference - while with autocompletion we have only prefix, here we have the whole symbol name so less "false positives" in the result so I thought I could leave it this way. In any case, I don't have a strong opinion about this and these two could indeed be unified. This function could go to tm_tag.c.